File: app/src/main/java/com/expensivebelly/dagger2retaingraph/activity/di/RetainModule.java
package com.expensivebelly.dagger2retaingraph.activity.di;

import com.expensivebelly.dagger2retaingraph.activity.RetainGraphPresenter;

import java.util.concurrent.TimeUnit;

import dagger.Module;
import dagger.Provides;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

@Module
class RetainModule {

    @Provides
    @ActivityScope
    static RetainGraphPresenter provideDaggerRetainGraphPresenter(Observable<String> timer) {
        return new RetainGraphPresenter(timer);
    }

    @Provides
    static Observable<String> provideObservable() {
        return Observable.interval(1, TimeUnit.SECONDS).flatMap(new Function<Long, ObservableSource<String>>() {
            @Override
            public ObservableSource<String> apply(Long aLong) throws Exception {
                return Observable.just(String.valueOf(aLong));
            }
        }).cache().subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/java/com/expensivebelly/dagger2retaingraph/activity/di/RetainModule.kt
package com.expensivebelly.dagger2retaingraph.activity.di

import com.expensivebelly.dagger2retaingraph.activity.MainPresenter
import dagger.Module
import dagger.Provides
import io.reactivex.rxjava3.core.Observable
import java.util.concurrent.TimeUnit

@Module
class RetainModule {

    @Provides
    @ActivityScope
    fun provideDaggerRetainGraphPresenter(timer: Observable<String>): MainPresenter {
        return MainPresenter(timer)
    }

    @Provides
    fun provideObservable(): Observable<String> {
        return Observable.interval(1, TimeUnit.SECONDS)
                .flatMap { aLong -> Observable.just(aLong.toString()) }
                .replay(1).refCount(1, TimeUnit.SECONDS)
    }
}